✨ Created DGD Fusion Annot WF

migbro opened the pull request:
<!--Pull Request Template-->

## Description

This creates a workflow to:
1. Updated fusion gene symbols
2. Format file
3. Annotated fusions

See README for details

Closes # (https://github.com/d3b-center/bixu-tracker/issues/1757)
Fixes # (https://github.com/PediatricOpenTargets/ticket-tracker/issues/490)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Various tool tests are here: https://cavatica.sbgenomics.com/u/d3b-bixu/d3b-dgd-collab-dev
- [ ] [Main workflow test](https://cavatica.sbgenomics.com/u/d3b-bixu/d3b-dgd-collab-dev/tasks/8bbd0f2e-02be-424c-9a64-427d6b5ada2f)

**Test Configuration**:
* Environment:
* Test files: Used s3://d3b-openaccess-us-east-1-prd-pbta/open-targets/v12/fusion-dgd.tsv.gz

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [x] I have committed any related changes to the PR

migbro committed:
:tada: initial commit

dmiller15 committed:
:sparkles: add gene symbol updater

migbro committed:
:tada: added some tools

migbro committed:
:tada: wrote fusion standardization script
:tada: wrote workflow
:hammer: refactored tool to use script

migbro committed:
:pencil: updated doc

dmiller15 commented on tools/update_gene_symbols.cwl:
```diff
@@ -0,0 +1,38 @@
+  updated_fusion_tsv:
+    type: File
+    outputBinding:
+      glob: '*.tsv*'
```
For simplicity.
```suggestion
      glob: $(inputs.output_filename)
```

dmiller15 commented on tools/update_gene_symbols.cwl:
```diff
@@ -0,0 +1,38 @@
+    inputBinding: {position:1, prefix: "--old_symbol"} }
+  new_symbol: { type: 'string?', doc: "Column name for the new gene symbol(s) in the HGNC TSV. Set to override script defaults",
+    inputBinding: {position:1, prefix: "--new_symbol"} }
+  update_columns: {type: 'string?', doc: "Space-separated column names from the Fusions TSV where to update gene names (e.g. -u foo bar blah). Set to override script defaults",
```
This should probably be a string array input.

dmiller15 commented on tools/update_gene_symbols.cwl:
```diff
@@ -0,0 +1,38 @@
+cwlVersion: v1.0
```
Let's do 1.2

dmiller15 commented on tools/fusion_standardization.cwl:
```diff
@@ -0,0 +1,35 @@
+cwlVersion: v1.0
```
1.2

dmiller15 commented on tools/fusion_annotator.cwl:
```diff
@@ -0,0 +1,36 @@
+cwlVersion: v1.0
```
1.2

dmiller15 commented on tools/update_gene_symbols.cwl:
```diff
@@ -0,0 +1,38 @@
+    ramMin: 4000
+    coresMin: 2
+  - class: DockerRequirement
+    dockerPull: 'pgc-images.sbgenomics.com/d3b-bixu/hotspots:0.1.0'
```
I think this will run just fine in any base python3 docker.

migbro replied:
Um, hmmm, I see, so you're saying just look for some public python3 docker? The only python-specific docker we have built other than that seems to be for v2.7 - I don't want to use that

dmiller15 replied:
Yeah you can use a public python docker.

dmiller15 commented on workflows/dgd_fusion_annotator_wf.cwl:
```diff
@@ -0,0 +1,70 @@
+  genome_tar: {type: 'File', doc: "STAR-Fusion CTAT Genome lib", "sbg:suggestedValue": {
+      class: File, path: 62853e7ad63f7c6d8d7ae5a8, name: GRCh38_v39_CTAT_lib_Mar242022.CUSTOM.tar.gz}}
+  genome_untar_path: {type: ['null', string], doc: "This is what the path will be when genome_tar is unpackaged", default: "GRCh38_v39_CTAT_lib_Mar242022.CUSTOM"}
+  col_num: {type: ['null', int], doc: "column number in file of fusion name, use 24 for arriba v1.1, 30 for v2, 1 for DGD", default: 1}
```
lets just go with `int?` notation.

migbro replied:
Ah, yeah, that was a bit of copy/paste since that tool is from the RNAseq wf...I was tempted to make the wf a sub module, but seems heavy-handed for just one tool. @sickler-alex also faced that issue...

dmiller15 commented on workflows/dgd_fusion_annotator_wf.cwl:
```diff
@@ -0,0 +1,70 @@
+  # annotate fusion
+  genome_tar: {type: 'File', doc: "STAR-Fusion CTAT Genome lib", "sbg:suggestedValue": {
+      class: File, path: 62853e7ad63f7c6d8d7ae5a8, name: GRCh38_v39_CTAT_lib_Mar242022.CUSTOM.tar.gz}}
+  genome_untar_path: {type: ['null', string], doc: "This is what the path will be when genome_tar is unpackaged", default: "GRCh38_v39_CTAT_lib_Mar242022.CUSTOM"}
```
`string?`

dmiller15 commented on workflows/dgd_fusion_annotator_wf.cwl:
```diff
@@ -0,0 +1,70 @@
+  sbg: https://sevenbridges.com
+
+hints:
+  - class: 'sbg:AWSInstanceType'
+    value: c5.2xlarge;ebs-gp2;400
```
Idk if it's wise to include some sort of reasoning for this choice. Kind of seems out of place.

migbro replied:
Sure, so, not a big deal at the moment, but it does shave cost for time storage. Time since it's a pretty small wf, can all run on same instance + faster cores. Cost with the time savings and the smaller disk footprint.

dmiller15 replied:
Yeah I think we talked about the reasoning for this elsewhere but if someone sees it and is like "this seems like a weirdly specific instance choice" it would be nice if we had a doc explaining why.

migbro replied:
Oddly enough - I was able to add a `doc:` to that field. I think that should do it?

dmiller15 replied:
Wild. Yeah, i guess that works.

dmiller15 reviewed (changes requested):
Some notes.

migbro committed:
Update tools/update_gene_symbols.cwl

Co-authored-by: Dan Miller <dmiller15@users.noreply.github.com>

migbro committed:
:muscle: Incorporated some PR suggestions from DM

migbro commented:
@dmiller15 implemented suggested, minor refactor test here: https://cavatica.sbgenomics.com/u/d3b-bixu/d3b-dgd-collab-dev/tasks/8bbd0f2e-02be-424c-9a64-427d6b5ada2f

migbro committed:
:pencil: updated doc for clarity

jharenza commented on README.md:
```diff
@@ -1,22 +1,31 @@
-# D3b Bixu Repository Template
+# D3b DGD Collaboration
 
-Use this template to bootstrap a new D3b bixu repository 
+Hosts various code and workflows of mutual interest to D3b and DGD 
```
just a general comment that this does not need to be specific to DGD, this can be used for any non-STAR-Fusion/non-arriba file

migbro replied:
Oh, well, I guess I can make that change. I was sort of cheating and making it more DGD-specific since that was the main customer. I had hid some related options, so I can bring that back in

jharenza commented on README.md:
```diff
@@ -1,22 +1,31 @@
 
-### Badges
+## DGD Fusion Annotation
+This is a workflow that annotates a DGD fusion file.
```
```suggestion
This is a workflow that annotates a custom fusion file.
```

jharenza commented on tools/fusion_annotator.cwl:
```diff
@@ -0,0 +1,36 @@
+  input_fusion_file: {type: File, doc: "Fusion file formatted by format_fusion_file.cwl, likely from arriba input"}
+  genome_tar: File
+  genome_untar_path: {type: ['null', string], doc: "This is what the path will be when genome_tar is unpackaged", default: "GRCh38_v39_CTAT_lib_Mar242022.CUSTOM"}
+  col_num: {type: ['null', int], doc: "column number in file of fusion name, use 24 for arriba v1.1, 30 for v2, 1 for DGD", default: 1}
```
seems more fullproof to use the column name?

migbro replied:
So, this is an outside package - that author made that choice :-/

jharenza replied:
oh i see

jharenza commented on tools/fusion_standardization.cwl:
```diff
@@ -0,0 +1,35 @@
+inputs:
+  fusions_tsv: { type: File, doc: "Input fusion TSV file, i.e. fusion-dgd.tsv.gz. Can be compressed or uncompressed",
+    inputBinding: {position: 1, prefix: "--fusions_tsv"} }
+  caller: { type: [{type: enum, name: caller, symbols: ["STARFUSION", "ARRIBA", "DGD", "CUSTOM"]}], doc: "Caller used to produce input",
```
```suggestion
  caller: { type: [{type: enum, name: caller, symbols: ["STARFUSION", "ARRIBA", "DGD", "CUSTOM"]}], doc: "Caller used to produce input",
```

aren't DGD and custom one in the same?

migbro replied:
So, they should be, but I think @sakshamphul , when he made this, was expecting the input to also have Gene2A and and Gene2B and annots already. I had thought that the behavior of adding the config file would be that it'd add those dummy columns if missing, instead he wrote them all as required. That's why, if you look at the R script, I pre-fill the DGD dataframe before passing to the function. We can either wait to fix that behavior now or later.

migbro committed:
:racehorse: updated default ref input
:pencil: minor param/readme changes

migbro committed:
:pause: good stopping point, added some "hidden" required cols
:tada: created annofuse script. needs opts

migbro committed:
:tada: added annofuse wrapper
:wrench: fixed col order in standardization
:hammer: added opts to annoFuse wrapper

migbro commented:
@jharenza , to kind of sum up why this isn't as generic as it could be:
1. The fusion_standardization function currently doesn't do enough to ensure complete downstream compatibility. Missing was `"LeftBreakpoint", "RightBreakpoint", JunctionReadCount', 'SpanningFragCount'`, none if which is in the DGD file. So in this wrapper script (a wrapper is needed at a minimum to read tsv into data frame), I fill in those gaps
2. for annoFuse, building on point 2, I had to dump the readthrough filter param for QC, and set the readcount filters to 0 as such info is not available
3. I guess that's it really

migbro commented:
Latest test task here: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/sd-bhjxbdqk-x01-rnaseq-analysis/tasks/c5273913-0d30-4e73-a9f2-2a6d803fb2b9/

migbro committed:
:pencil: updated readme

jharenza commented:
> Latest test task here: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/sd-bhjxbdqk-x01-rnaseq-analysis/tasks/c5273913-0d30-4e73-a9f2-2a6d803fb2b9/

this is empty?

migbro commented:
@jharenza whoops! That was that task that Bo was working on. This is the real deal: https://cavatica.sbgenomics.com/u/d3b-bixu/d3b-dgd-collab-dev/tasks/ae601c7f-576d-4cb1-8407-52ef3a2f6aa9/

jharenza commented:
ALmost perfect - can we update `reciprocal_exists` to NA

migbro commented:
@jharenza  So, I can set this flag `checkReciprocal = FALSE` when running annoFuse. The effect of that seems to be that the column will be skipped. Is that ok, or would you like me to add it as a dummy col with `NA` in the end?

jharenza commented:
skipping is OK!

migbro commented:
I forgot, lastly, do you want to keep both outputs, or just final annoFuse?

jharenza commented:
final annofuse is fine

migbro committed:
:broom: finishing touches from PR suggestion

migbro commented:
Ok, great! Just approve the PR and I can merge it :)

jharenza reviewed (approved):
🚀
